### PR TITLE
revoke promise when provider is gone

### PR DESCRIPTION
**Pull request by zizhong**
This is to fix the memory leak issue #4137 .

### CODE REVIEW

**Comment by zizhong** [comment]
better to backport this if needed.

**Comment by zizhong** [comment]
kindly ping if you got time. @bryancall @zwoop @SolidWallOfCode @shinrich @scw00 @oknet

File: lib/cppapi/include/atscppapi/Async.h [hunk 1]
```diff
@@ -209,13 +213,20 @@ template <typename AsyncProviderType> class AsyncReceiver : noncopyable
...
+    receiver_promises_.erase(dispatch_controller_ptr);
+  }
 
 protected:
   AsyncReceiver() {}
   friend class Async;
 
 private:
-  mutable std::list<std::shared_ptr<AsyncReceiverPromise<AsyncReceiver<AsyncProviderType>, AsyncProviderType>>> receiver_promises_;
+  mutable std::unordered_map<AsyncDispatchController<AsyncReceiver<AsyncProviderType>, AsyncProviderType> *,
```
Comment (SolidWallOfCode) [inline lib/cppapi/include/atscppapi/Async.h]: Do you need `unordered_map`? Could you use `unordered_set` with a specialized hash that extracts the raw pointer from the `shared_ptr`?

**Comment by zizhong** [comment]
@shinrich @SolidWallOfCode list header is removed.

**Comment by zizhong** [comment]
Conflicts resolved.

**Comment by zizhong** [comment]
@shinrich @SolidWallOfCode mind take another look? Thank you very much!

**Pull request merged**